File: test-suite-java/src/test/java/io/micronaut/interceptor/test/state/CountingInterceptor.java
package io.micronaut.interceptor.test.state;

import jakarta.interceptor.AroundInvoke;
import jakarta.interceptor.Interceptor;
import jakarta.interceptor.InvocationContext;

@Interceptor
@Counted
public class CountingInterceptor {

    private int count;

    @AroundInvoke
    public Object count(InvocationContext context) throws Exception {
        count++;
        CountingInterceptors.seen(context.getTarget(), this);
        return context.proceed();
    }

    public int count() {
        return count;
    }
}

File: test-suite-kotlin/src/test/kotlin/io/micronaut/interceptor/test/CountingInterceptor.kt
package io.micronaut.interceptor.test

import jakarta.interceptor.AroundInvoke
import jakarta.interceptor.InvocationContext

/**
 * A plain class named directly by `@Interceptors`, without `@Interceptor`.
 */
open class CountingInterceptor {

    @AroundInvoke
    open fun count(context: InvocationContext): Any? {
        Calls.recorded += "count"
        return context.proceed()
    }
}
